File: tests/fixtures/typescript/filters.ts
// Filters.ts — transformer implementations

import { Transformer } from "./processor";

export class NormalizeFilter implements Transformer {
  transform(data: Record<string, unknown>): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(data)) {
      result[k] = typeof v === "string" ? v.trim().toLowerCase() : v;
    }
    return result;
  }
}

export class RemoveNullFilter implements Transformer {
  transform(data: Record<string, unknown>): Record<string, unknown> {
    const result: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(data)) {
      if (v !== null && v !== undefined) {
        result[k] = v;
      }
    }
    return result;
  }
}

export function chainFilters(
  ...filters: Transformer[]
): Transformer {
  return {
    transform(data: Record<string, unknown>): Record<string, unknown> {
      let r = data;
      for (const f of filters) {
        r = f.transform(r);
      }
      return r;
    },
  };
}

File: tests/fixtures/go/filters.go
// filters.go — transformer implementations

package fixtures

import "strings"

// NormalizeFilter strips and lowercases string values
type NormalizeFilter struct{}

func (f NormalizeFilter) Transform(data map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range data {
		if s, ok := v.(string); ok {
			result[k] = strings.ToLower(strings.TrimSpace(s))
		} else {
			result[k] = v
		}
	}
	return result
}

// RemoveNullFilter removes nil values
type RemoveNullFilter struct{}

func (f RemoveNullFilter) Transform(data map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range data {
		if v != nil {
			result[k] = v
		}
	}
	return result
}

// ChainFilters composes multiple transformers into one
func ChainFilters(filters ...Transformer) Transformer {
	return &chainTransformer{filters: filters}
}

type chainTransformer struct {
	filters []Transformer
}

func (c *chainTransformer) Transform(data map[string]interface{}) map[string]interface{} {
	r := data
	for _, f := range c.filters {
		r = f.Transform(r)
	}
	return r
}
